jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -0,0 +1,339 @@
+
+  /**
+   * Bakery client inspired by the equivalent GO code.
+   *
+   * This object exposes the ability to perform requests
+   * that automatically acquire and discharge macaroons
+   *
+   * @class Bakery
+   */
+
+  var Bakery = Y.Base.create('Bakery',
+      Y.Base, [], {
+
+      /**
+       Initialize.
+
+       @method initializer
+       @return {undefined} Nothing.
+       */
+      initializer: function () {
+        this.webhandler = new Y.juju.environments.web.WebHandler();
+      },
+
+      /**
+       Send get request.
+
+       @param {String} The path to make the api request to.
+       @param {String} The path to make a request to set the cookie.
+       @param {Function} successCallback Called when the api request completes
+       successfully.
+       @param {Function} failureCallback Called when the api request fails
+       with a response of >= 400 except 401/407 where it does authentication.
+       */
+      sendGetRequest: function (path, setCookiePath,
+                                successCallback, failureCallback) {
+        var macaroons = Y.Cookie.get('Macaroons');
+        var headers = {'Bakery-Protocol-Version': 1};
+        if (macaroons !== null) {
+          headers['Macaroons'] = macaroons;
+        }
+
+        this.webhandler.sendGetRequest(
+          path, headers, null, null, false, null,
+          this._requestHandlerWithInteraction.bind(
+            this,
+            path,
+            setCookiePath,
+            successCallback,
+            failureCallback
+          )
+        );
+      },
+
+      /**
+       Handles the request response from the _makeRequest method, calling the
+       supplied failure callback if the response status was >= 400 or passing
+       the response object to the supplied success callback. For 407/401
+       response it will request authentication through the macaroon provided in
+       the 401/407 response.
+
+       @method _requestHandlerWithInteraction
+       @param {String} The path to make the api request to.
+       @param {String} The path to make a request to set the cookie.
+       @param {Function} successCallback Called when the api request completes
+       successfully.
+       @param {Function} failureCallback Called when the api request fails
+       with a response of >= 400 (except 401/407).
+       @param {Object} response The XHR response object.
+       */
+      _requestHandlerWithInteraction: function (path,
+                                                setCookiePath,
+                                                successCallback,
+                                                failureCallback,
+                                                response) {
+        var target = response.target;
+        if (target.status === 401 &&
+          target.getResponseHeader('Www-Authenticate') === 'Macaroon') {
+          var jsonResponse = JSON.parse(target.responseText);
+          this._authenticate(
```

Comment:
There are a lot of methods being passed around here. What might be a little cleaner is setting a depth !> 1 requirement. So you instead of passing in the `_sendOriginalRequest` to the `_authenticate` method you would pass the necessary arguments for all subsequent calls to the `_authenticate` method. I don't see that `_authenticate` is being called anywhere else.